File: es/codes/javascript/chapter_backtracking/permutations_i.js
/**
 * File: permutations_i.js
 * Created Time: 2023-05-13
 * Author: Justin (xiefahit@gmail.com)
 */

/* 回溯算法：全排列 I */
function backtrack(state, choices, selected, res) {
    // 当状态长度等于元素数量时，记录解
    if (state.length === choices.length) {
        res.push([...state]);
        return;
    }
    // 遍历所有选择
    choices.forEach((choice, i) => {
        // 剪枝：不允许重复选择元素
        if (!selected[i]) {
            // 尝试：做出选择，更新状态
            selected[i] = true;
            state.push(choice);
            // 进行下一轮选择
            backtrack(state, choices, selected, res);
            // 回退：撤销选择，恢复到之前的状态
            selected[i] = false;
            state.pop();
        }
    });
}

/* 全排列 I */
function permutationsI(nums) {
    const res = [];
    backtrack([], nums, Array(nums.length).fill(false), res);
    return res;
}

// Driver Code
const nums = [1, 2, 3];
const res = permutationsI(nums);

console.log(`输入数组 nums = ${JSON.stringify(nums)}`);
console.log(`所有排列 res = ${JSON.stringify(res)}`);

File: ja/codes/ruby/chapter_backtracking/permutations_i.rb
=begin
File: permutations_i.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### バックトラッキング法：全順列 I ###
def backtrack(state, choices, selected, res)
  # 状態の長さが要素数に等しければ、解を記録
  if state.length == choices.length
    res << state.dup
    return
  end

  # すべての選択肢を走査
  choices.each_with_index do |choice, i|
    # 枝刈り：要素の重複選択を許可しない
    unless selected[i]
      # 試行: 選択を行い、状態を更新
      selected[i] = true
      state << choice
      # 次の選択へ進む
      backtrack(state, choices, selected, res)
      # バックトラック：選択を取り消し、前の状態に戻す
      selected[i] = false
      state.pop
    end
  end
end

### 全順列 I ###
def permutations_i(nums)
  res = []
  backtrack([], nums, Array.new(nums.length, false), res)
  res
end

### Driver Code ###
if __FILE__ == $0
  nums = [1, 2, 3]

  res = permutations_i(nums)

  puts "入力配列 nums = #{nums}"
  puts "すべての順列 res = #{res}"
end
